domain/secret/state/state.go

```diff
@@ -1883,3 +1949,268 @@ ON CONFLICT(secret_id) DO UPDATE SET
+			}
+		}
+		perm.SubjectUUID = result.UUID
+		perm.SubjectTypeID = params.SubjectTypeID
+
+		// Look up the UUID of the access scope entity.
+		result = entityRef{}
+		err = tx.Query(ctx, selectScopeUUIDStmt, selectScopeQueryParams).Get(&result)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return fmt.Errorf("%#v %q not found%w", params.ScopeTypeID, params.ScopeID, errors.Hide(scopeNotFoundError))
+			} else {
+				scope := params.ScopeID
+				if params.ScopeTypeID == domainsecret.ScopeModel {
+					scope = "model"
+				}
+				return errors.Annotatef(err, "looking up secret grant scope UUID for %q", scope)
+			}
+		}
+		perm.ScopeUUID = result.UUID
+		perm.ScopeTypeID = params.ScopeTypeID
+
+		// Check that the access scope or subject type is not changing.
+		id := secretID{}
+		err = tx.Query(ctx, checkInvariantStmt, perm).Get(&id)
+		if err == nil {
+			// Should never happen.
+			return secreterrors.InvalidSecretPermissionChange
+		} else if !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotatef(err, "checking duplicate permission record for secret %q", uri)
+		}
+
+		return st.grantAccess(ctx, tx, perm)
+	})
+	return errors.Trace(domain.CoerceError(err))
+}
+
+func (st State) grantAccess(ctx context.Context, tx *sqlair.TX, perm secretPermission) error {
+	insertQuery := `
+INSERT INTO secret_permission (*)
+VALUES ($secretPermission.*)
+ON CONFLICT(secret_id, subject_uuid) DO UPDATE SET
+    role_id=excluded.role_id,
+    -- These are needed to fire the immutable trigger.
+    subject_type_id=excluded.subject_type_id,
+    scope_type_id=excluded.scope_type_id,
+    scope_uuid=excluded.scope_uuid
+`
+
+	insertStmt, err := st.Prepare(insertQuery, secretPermission{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := tx.Query(ctx, insertStmt, perm).Run(); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+
+}
+
+// GetSecretAccess returns the access to the secret for the specified accessor.
+// It returns an error satisfying [secreterrors.SecretNotFound] if the secret is not found.
+func (st State) GetSecretAccess(ctx context.Context, uri *coresecrets.URI, params domainsecret.AccessParams) (string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	query := `
+SELECT sr.role AS &M.role
+FROM   v_secret_permission sp
+JOIN   secret_role sr ON sr.id = sp.role_id
+WHERE  secret_id = $secretAccessor.secret_id
+AND    subject_type_id = $secretAccessor.subject_type_id
+AND    subject_id = $secretAccessor.subject_id
+`
+
+	selectStmt, err := st.Prepare(query, secretAccessor{}, sqlair.M{})
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	access := secretAccessor{
+		SecretID:      uri.ID,
+		SubjectTypeID: params.SubjectTypeID,
+		SubjectID:     params.SubjectID,
+	}
+	var role string
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		if isLocal, err := st.checkExistsIfLocal(ctx, tx, uri); err != nil {
```

Comment:
I see this in many different places, checkExistsIfLocal really should raise the `secreterrors.SecretIsNotLocal` error if it's not local.
Otherwise, we should rename the method to `getIfLocal`.

Reply:
In 2 places (Get and Save secret consumer) we do want to handle both the local and remote case. In the other call sites, we only expect a local URI and the error is a defensive fallback.
The method is called checkExistsIfLocal because if the secret is local, we do want to return a SecretNotFound error if it doesn't exist. Changing the method name would obscure this behaviour.